### PR TITLE
make restoration work

### PR DESCRIPTION
**Pull request by goderbauer**

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import 'package:flutter/material.dart';
 
 void main() {
-  runApp(BooksApp());
+  runApp(RootRestorationScope(
```
Comment (goderbauer) [inline lib/main.dart:4]: We need to manually create a root scope (usually this is created by MaterialApp, but we have state to restore in BooksApp which is an ancestor of MaterialApp and doesn't have access to the root scope created by it).

File: lib/main.dart [hunk 2]
```diff
@@ -16,18 +19,26 @@ class BooksApp extends StatefulWidget {
   State<StatefulWidget> createState() => _BooksAppState();
 }
 
-class _BooksAppState extends State<BooksApp> {
-  BookRouterDelegate _routerDelegate = BookRouterDelegate();
+class _BooksAppState extends State<BooksApp> with RestorationMixin {
+  RestorableBookDelegate _routerDelegate = RestorableBookDelegate();
```
Comment (goderbauer) [inline lib/main.dart:23]: The custom RestorableProperty "RestorableBookDelegate" is a wrapper around BookRouterDelegate and responsible for restoring the delegate's state.

**Comment by goderbauer** [comment]
There appears to be one annoyances: `RouterDelegate.setInitialRoutePath` (which by default just calls `RouterDelegate.setNewRoutePath`) is called during state restoration with an incorrect path (it's the path the app was initially launched with). For this particular example this doesn't seem to matter because that call gets ignored. However, for other use cases that could mean the delegate will configure itself to the path the the app was initially launched with overwritten the path restored during state restoration... 

Potential solutions out of this:
* Router doesn't call `RouterDelegate.setInitialRoutePath` during state restoration and the delegate has to restore its state on its own (as implemented in this PR)
* When state is collected prior to an app getting killed, the Router could actually persist the current path the RouterDelegate is configured to (with the help of `RouterDelegate.currentConfiguration` and the `BookRouteInformationParser`). During state restoration, that path is then handed back to the delegate via the `RouterDelegate.setInitialRoutePath` call. The delegate then configures its internal state based on that. It would not need to implement anything special for state restoration and the changes in this PR become moot.

Now that I think about it, the second approach actually sounds like a better user experience? Let me file an issue for this to discuss this further...

**Comment by goderbauer** [comment]
Filed https://github.com/flutter/flutter/issues/81231. I think we can make this simpler.

**Comment by justinmc** [comment]
Thanks for this and for all of the explanation!  I commented on your issue.

**Pull request merged**